gmt-4.5.14/src/spotter/rotconverter.1
.TH ROTCONVERTER 1 "1 Nov 2015" "GMT 4.5.14" "Generic Mapping Tools"
.SH NAME
rotconverter \- Manipulate finite and stage rotations 
.SH SYNOPSIS
\fBrotconverter\fP [ \fB+\fP|\fB-\fP ] \fIrotA\fP [ \fB+\fP|\fB-\fP \fIrotB\fP ] [ \fB+\fP|\fB-\fP \fIrotC\fP ] ... 
[ \fB\-C\fP[\fBa\fP|\fBt\fP]] [ \fB\-D\fP ] [ \fB\-E\fP[\fIfact\fP] ] [ \fB\-F\fP\fIin\fP/\fPout\fP ] [ \fB\-N\fP ] [ \fB\-S\fP ] 
[ \fB\-T\fP ] [ \fB\-V\fP ]
.SH DESCRIPTION
\fBrotconverter\fP reads one or more plate motion models (rotations) stored in the
given files.  If more than one plate motion model is given we will add or subtract
them in the order they were listed.  The minus sign means we should first transpose the
rotation and then add it to the previous rotation.  The input files must all be of the
same type (stage poles or finite rotations) which may differ from the desired output
format; see \fB\-F\fP.  If a file cannot be opened we will attempt to decode the file name
as a single rotation whose parameters are separated by slashes.
.br
	No space between the option flag and the associated arguments.  Use upper case for the
option flags and lower case for modifiers.
.TP
\fIrotX\fP
Name of a file with a plate motion model.  Separate several files with desired operator
(+ or -).  The very first file may also have a leading minus to imply a transpose.
If any of the specified rotation models cannot be opened as a file, we will try to decode
the file name as \fIlon/lat/tstart\fP[\fI/tstop\fP]/\fIangle\fP for a single rotation given
on the command line.  The \fItstop\fP argument is required for stage poles only.  For a
single finite rotation without any time information, give \fIlon/lat/angle\fP only.
.SH OPTIONS
.TP
\fB\-C\fP
Write out a column header record identifying the various columns [Default is no header record].
Append \fBa\fP to indicate opening angles and \fBt\fP to indicate opening rates [Default].
.TP
\fB\-D\fP
Report longitudes use the -180/+180 range [Default is 0/360].
.TP
\fB\-E\fP
Scale opening angles by \fIfact\fP on output.  Requires stage pole output (see \fB\-F\fP).
.TP
\fB\-F\fP
Specify both the input and output format for rotations.  The \fIin\fP and \fIout\fP flags
must be either \fBf\fP or \fBs\fP for finite or stage rotations, respectively.  Note that
both must be specified if \fB\-F\fP is set [Default is \fB\-Fff\fP (both input and output
are finite rotations)].
.TP
\fB\-N\fP
Place all output poles in the northern hemisphere [Default reports positive rotation angles].
.TP
\fB\-S\fP
Place all output poles in the southern hemisphere [Default reports positive rotation angles].
.TP
\fB\-T\fP
Transpose the final result, i.e., change the sign of the rotation angles.
.TP
\fB\-V\fP
Selects verbose mode, which will send progress reports to stderr [Default runs "silently"].
Report statistics of extracted rotations.
.SH EXAMPLES
To convert the finite rotations in the file model_finite.APM to stage poles, run
.br
.sp
\fBrotconverter\fP model_finite.APM \fB\-Ffs\fP > model_stages.APM
.br
.sp
To obtain Nazca motion relative to Pacific hotspots by adding the motion
of Nazca relative to a fixed Pacific to the Pacific-Hotspot reference model DC85_stages.d,
and report the result as finite reconstruction poles in the northern hemisphere, try
.br
.sp
\fBrotconverter\fP DC85_stages.APM + Pac_Naz_stages.RPM \fB\-N\fP \fB\-Fsf\fP > Naz_HS_finite.APM
.br
.sp
To add the final rotations ROT(150.1, 70.5, -20.3) and ROT (145.0, 40.0, 11.4),  try
.br
.sp
\fBrotconverter\fP 150.1/70.5/-20.3 + 145/40/11.4
.sp
.br
which prints out 157.32, -80.44, 11.97.
.br
.sp
To make stage rotations suitable for generating flowlines (fracture zones) from a model
of relative plate motions PL1-PL2.RPM, assuming symmetric spreading,, try
.br
.sp
\fBrotconverter\fP PL1-PL2.RPM \fB\-E\fP \fB\-Ffs\fP > PL1-PL2_half.RPM
.br
\fBrotconverter\fP - PL1-PL2.RPM \fB\-E\fP \fB\-Ffs\fP > PL2-PL1_half.RPM
.SH "SEE ALSO"
.IR backtracker (1),
.IR grdrotater (1),
.IR grdspotter (1),
.IR hotspotter (1),
.IR originator (1)


gmt-4.5.14/ChangeLog
-----------------------------------------------------------------------------------------
	$Id: ChangeLog 10331 2015-10-29 19:41:56Z pwessel $
-----------------------------------------------------------------------------------------

-----------------------------------------------------------------------------------------
2015-11-01	pwessel
	* GMT 4.5.14 finalized and released
-----------------------------------------------------------------------------------------
2015-10-29	pwessel
	* meca/psmeca.c:	Enforce rakes and strikes to be in -180/+180 range.
				Issue #799, for gmt4 only as not a problem in gmt5.

2015-07-28	pwessel
	* meca/psmeca.c:	Be able to read any value x and y coordinates

2015-06-24	pwessel
	* gmt_support.c:	Fix bug in jump near Dateline in gmt_inonout_sphpol_count,
				affecting inside/outside spherical cases, reported in message-2219

2015-03-18	pwessel
	* share/cpt/GMT_wysiwyg.cpt:	Fix typo in blue for high end [Issue #689].

2015-01-16	pwessel
	* gmt_plot.c:		Fix issue # 662 with graph arrows for negative scales.

2015-01-10	pwessel
	* gmt_plot.c:		Flipping the n/s sides array when lon and -y axes
				fixes issue #520

2015-01-01	pwessel
	* configure.ac:		Moved to 4.5.14 and back to svn release only.
-----------------------------------------------------------------------------------------
2015-01-01	pwessel
	* GMT 4.5.13 finalized and released [GSHHG 2.3.4 was released on Jan 1 as well]
-----------------------------------------------------------------------------------------
2014-12-28	pwessel
	* gmt_map.c:		Needed separate GMT_cartesian_overlap for non-geographic
				projections.  Fixes issue # 657.

2014-09-29	pwessel
	* gmt_plot.c:		Left a PSL variable undefined when grdimage plotted a grid
				with x = longitude and y = Cartesian.

2014-09-09	pwessel
	* grdfilter.c:		Fix y-shifts when -I sets smaller output.  Issue #616.
	* gmt_bcr.c:		Fix problem with points slightly outside causing an
				access violation; see issue #620.

2014-08-29	pwessel
	* sample1d.c:		Fix resampling for decreasing t values.  Issue #608.

2014-08-05	pwessel
	* gmt_map.c:		Could run into trouble clipping polygons for non-periodic
				maps.  Added more checks in GMT_wesn_clip.

2014-07-31	pwessel
	* gmt_map.c:		Rect clipping with -R<w/s/e/n>r could introduce stray lines
				when dataset has features about 180 degrees away from
				the user's area. See issue # 594.

2014-07-30	pwessel
	* configure.ac:		Set default GSHHG version to 2.3.2 (released August 1, 2014)
	* psxy[z].c:		Fixed wrap-around issue in fault lines.

2014-07-27	pwessel
	* gmtselect.c:		The -Z option needed to pass records with NaNs.

2014-07-19	pwessel
	* x2sys/x2sys_solve.c:	Fix problems that arise when a track has less crossovers
				than model parameters.
	* gmtmath.c:		Fix issue #123 with files and -C.

2014-07-01	pwessel
	* configure.ac:		Set default GSHHG version to 2.3.1 (released July 1, 2014)
	
2014-06-28	pwessel
	* mgd77/mgd77sniffer.c:	Needed to check that grid values used in decimate function
	 			also were within reasonable range.

2014-05-26	pwessel
	* gmt_map.c:		The map region clip path for -JE/e was hardwired to a full
				circle and did not care about -R setting.
				When -R...r and -Je/E was given the search for enclosing
				geographic boundaries failed when either the S or N pole
				was the projection's antipole.
2014-05-25	pwessel
	* grdmath.c:		Fix issue #566 bug in grdmath's KM2DEG operator.

2014-03-19	pwessel
	* gmtselect.c:		The -Fpolygon option might miss points because the io-machinery
				would sometimes set the min/max longitudes found incorrectly.

2014-03-14	pwessel
	* gmt_shore.c:		Give more error feedback if nc_open fails due to old library.

2014-03-01	pwessel
	* configure.ac:		Moved to 4.5.13 and back to svn release only.

-----------------------------------------------------------------------------------------
2014-03-01	pwessel
	* GMT 4.5.12 finalized and released [GSHHG 2.3.0 was released on Feb 1]
-----------------------------------------------------------------------------------------
2014-02-17	fwobbe
	* gmt_init.c etc.:	File locking -- if enabled via -DFLOCK -- is
				non-compulsory now.  This means if fcntl fails for some reason,
				GMT continues to access the file.  Hence, it is recommended to
				configure GMT with --enable-flock even if fcntl might be buggy
				on some ancient NFS shares.

2014-02-17	pwessel
	* gmt_map.c:		Size for memory allocation cast the product of nx * ny to
				size_t but overflow had already happen as calculation was
				done using ints, then cast to size_t. Must cast nx and ny
				separately.
2014-02-10	jluis
	* misc/gmt2kml.c:	Make sure the default style IDs are unique for each process.

2014-02-07	jluis
	* ps2raster.c:		Make sure that images in KML files do not carry a directory name

2014-02-03	pwessel
	* pslib.c:		Auto-widen paragraph width to at least fit widest word.

2014-02-01	pwessel
	* configure.ac:		Since GSHHG 2.3.0 was released today, make that the default

2014-01-26	pwessel
	* gmt_grdio.c:		Building the grid command string could exceed its length if
				the argument became exactly the max length of the array; then
				a final extra character \0 would be added and exceed array length.
				Thanks to Joachim Saul for pointing it out.

2014-01-22	pwessel
	* pscoast.c:		The processing of -I -N levels and pens could get out of sync
				so that the wrong pen was used for the specified feature.

2014-01-16	remko
	* pslib.c:		Take superscripts and subscripts into account when determining
				dimensions of text box.

2014-01-10	pwessel
	* grdmath.c:		Had to let SDIST return km instead of degrees since it cannot
				do degrees for geodesics.  By doing km it is now aligned with
				what PDIST and LDIST returns for geographic data, and matches
				what GMT5 does.  Some scripts had to be modified to handle the
				new behavior.  If you use SDIST you may wish to check your usage.
				Also fixed bug in SDIST in that spherical selected geodesics

gmt-4.5.14/src/spotter/rotconverter.c
/*--------------------------------------------------------------------
 *	$Id: rotconverter.c 10289 2014-12-28 21:17:06Z pwessel $
 *
 *   Copyright (c) 1999-2015 by P. Wessel
 *
 *   This program is free software; you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License as published by
 *   the Free Software Foundation; version 2 or any later version.
 *
 *   This program is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   Contact info: www.soest.hawaii.edu/pwessel
 *--------------------------------------------------------------------*/
/*
 * Program for converting between finite and stage poles or to add rotations.
 *
 * Author:	Paul Wessel, SOEST, Univ. of Hawaii, Honolulu, HI, USA
 * Date:	24-OCT-2001
 * Version:	1.0
 *		31-MAR-2006: Changed -H to -C to avoid clash with GMT -H
 *
 *-------------------------------------------------------------------------
 * An ASCII stage pole (Euler) file must have following format:
 *
 * 1. Any number of comment lines starting with # in first column
 * 2. Any number of blank lines (just carriage return, no spaces)
 * 2. Any number of stage pole records which each have the format:
 *    lon(deg)  lat(deg)  tstart(Ma)  tstop(Ma)  ccw-angle(deg)
 * 3. stage records must go from oldest to youngest rotation
 * 4. Note tstart is larger (older) that tstop for each record
 * 5. No gaps allowed: tstart must equal the previous records tstop
 *
 * Example: Duncan & Clague [1985] Pacific-Hotspot rotations:
 *
 * # Time in Ma, angles in degrees
 * # lon  lat	tstart	tend	ccw-angle
 * 165     85	150	100	24.0
 * 284     36	100	74	15.0
 * 265     22	74	65	7.5
 * 253     17	65	42	14.0
 * 285     68	42	0	34.0
 *
 * AN ASCII finite pole must have the following format:
*
 * 1. Any number of comment lines starting with # in first column
 * 2. Any number of blank lines (just carriage return, no spaces)
 * 2. Any number of finite pole records which each have the format:
 *    lon(deg)  lat(deg)  tstop(Ma)  ccw-angle(deg)
 * 3. finite rotations must go from youngest to oldest rotation
 *
 * Example: Duncan & Clague [1985] Pacific-Hotspot rotations:
 *
 * # Time in Ma, angles in degrees
 * #longitude	latitude	time(My)	angle(deg)
 * 285.00000	 68.00000	 42.0000	 34.0000
 * 275.66205	 53.05082	 65.0000	 43.5361
 * 276.02501	 48.34232	 74.0000	 50.0405
 * 279.86436	 46.30610	100.0000	 64.7066
 * 265.37800	 55.69932	150.0000	 82.9957
  */

#include "spotter.h"

int main (int argc, char **argv)
{
	struct EULER *p = NULL;		/* Pointer to array of stage poles */
	struct EULER *a = NULL, *b = NULL;		/* Pointer to arrays of stage poles */

	GMT_LONG i, j, k, n_p, n_a = 1, n_b;	/* Misc. counters */
	GMT_LONG last_sign, n_slash;
	GMT_LONG rate_or_rad = 0;		/* Default unit is degrees/My */

	GMT_LONG error = FALSE;		/* Set to TRUE if arguments are inconsistent */
	GMT_LONG geodetic = TRUE;	/* Want 0-360 range */
	GMT_LONG finite_in = TRUE;	/* Default format is finite rotation poles for both input and output */
	GMT_LONG finite_out = TRUE;	/* FALSE gives stage poles */
	GMT_LONG first = TRUE;		/* TRUE for first input file */
	GMT_LONG south = FALSE;		/* TRUE if we want to report poles in southern hemisphere */
	GMT_LONG north = FALSE;		/* TRUE if we want to report poles in northern hemisphere */
	GMT_LONG online_rot = FALSE;	/* TRUE if we gave a rotation on the commandline rather than file name */
	GMT_LONG no_time = FALSE;	/* TRUE if we gave a rotation on the commandline as lon/lat/angle only */
	GMT_LONG header = FALSE;		/* TRUE to write out a header record */
	GMT_LONG transpose = FALSE;	/* TRUE if we want to change the sign of the final rotation */
	GMT_LONG reduce_angle = FALSE;	/* TRUE to scale stage pole angles by reduce_fact */

	double zero = 0.0;		/* Needed to pass to spotter_init */
	double lon, lat;		/* Pole location for online rotations */
	double t0, t1;			/* Start, stop times for online rotations */
	double angle;			/* Rotation angle for online rotations */
	double reduce_fact = 0.5;	/* To get half-angles */

	char *start_text[2] = {"tstart(My)", "astart(deg)"};	/* Misc. column titles for rates or angles */
	char *end_text[2] = {"tend(My)", "aend(deg)"};
	char *time_text[2] = {"ttime(My)", "tangle(deg)"};
	char fmt[BUFSIZ];

	argc = (int)GMT_begin (argc, argv);

	for (i = 1; i < argc; i++) {
		if (argv[i][0] == '-' && argv[i][1]) {
			switch (argv[i][1]) {

				/* Common parameters */

				case 'V':
				case ':':
					error += GMT_parse_common_options (argv[i], NULL, NULL, NULL, NULL);
					break;

				/* Supplemental parameters */

				case 'F':
					if (strlen (argv[i]) != 4) {
						fprintf (stderr, "%s: ERROR: Must specify -F<in><out>\n", GMT_program);
						error++;
						continue;
					}
					switch (argv[i][2]) {	/* Input/output format */
						case 'F':	/* Finite rotations */
						case 'f':
							finite_in = TRUE;
							break;
						case 'S':	/* Stage rotations */
						case 's':
							finite_in = FALSE;
							break;
						default:
							fprintf (stderr, "%s: ERROR: Must specify f|s\n", GMT_program);
							error++;
							break;
					}
					switch (argv[i][3]) {	/* Output format */
						case 'F':	/* Finite rotations */
						case 'f':
							finite_out = TRUE;
							break;
						case 'S':	/* Stage rotations */
						case 's':
							finite_out = FALSE;
							break;
						default:
							fprintf (stderr, "%s: ERROR: Must specify f|s\n", GMT_program);
							error++;
							break;
					}
					break;

				case 'C':	/* Write column header record */
					header = TRUE;
					if (argv[i][2] == 'a') rate_or_rad = 1;
					break;

				case 'D':	/* Convert to finite rotation poles instead */
					geodetic = FALSE;
					break;

				case 'E':	/* Convert to finite rotation poles instead */
					reduce_angle = TRUE;
					if (argv[i][2]) reduce_fact = atof (&argv[i][2]);
					break;

				case 'N':	/* Ensure all poles reported are in northern hemisphere */
					north = TRUE;
					break;

				case 'S':	/* Ensure all poles reported are in southern hemisphere */
					south = TRUE;
					break;

				case 'T':	/* Transpose the final result (i.e., change sign of rotation) */
					transpose = TRUE;
					break;
					
				case '0': case '1': case '2': case '3': case '4': case '5': case '6':
					case '7': case '8': case '9': case '.':
					break;	/* Probably a rotation lon/lat/angle with negative longitude */
				default:					
					error = TRUE;
					GMT_default_error (argv[i][1]);
					break;
			}
		}
	}

	if (argc == 1 || GMT_give_synopsis_and_exit) {
		fprintf (stderr, "%s %s - Manipulate finite and stage rotations\n\n", GMT_program, SPOTTER_VERSION);
		fprintf (stderr, "usage: %s [+][-] rotA [[+][-] rotB] [[+][-] rotC] ... [-C[a|t]] [-D] [-E[<factor>]] [-F<in><out>]\n", GMT_program);
		fprintf (stderr, "\t[-N] [-S] [-T] [-V] > outfile\n\n");
		if (GMT_give_synopsis_and_exit) exit (EXIT_FAILURE);

		fprintf (stderr, "\trotA, rotB, etc. are finite or stage rotation pole files (see -F to set which kind).\n");
		fprintf (stderr, "\t   Alternatively, they can be a single rotation in lon/lat[/tstart[/tstop]]/angle format.\n");
		fprintf (stderr, "\t   For finite rotations with no time info, give lon/lat/angle only.\n");
		fprintf (stderr, "\t   The rotations will be added/subtracted in the order given.\n");
		fprintf (stderr, "\tOPTIONS:\n\n");
		fprintf (stderr, "\t-C Write a column header record with names of each column [No header].\n");
		fprintf (stderr, "\t   Append a if opening angles and t if opening rates [Default]\n");
		fprintf (stderr, "\t-D Report longitudes in -180/+180 range [ Default is 0-360]\n");
		fprintf (stderr, "\t-E Reduce opening angles for stage rotations by factor [0.5]\n");
		fprintf (stderr, "\t   Typically used to get half-rates needed for flowlines\n");
		fprintf (stderr, "\t-F Set input and output file type: f for finite and s for stage rotations [Default is -Fff]\n");
		fprintf (stderr, "\t-N Ensure all poles are in northern hemisphere [ Default ensures positive opening angles/rates]\n");
		fprintf (stderr, "\t-S Ensure all poles are in southern hemisphere [ Default ensures positive opening angles/rates]\n");
		fprintf (stderr, "\t-T Transpose the result (i.e., change sign of rotation angle)\n");
		GMT_explain_option ('V');
		exit (EXIT_FAILURE);
	}

	if (south && north) {
		fprintf (stderr, "%s: ERROR: Cannot specify both -N and -S!\n", GMT_program);
		error++;
	}
	if (reduce_angle && finite_out) {
		fprintf (stderr, "%s: ERROR: -E requires stage rotations on output\n", GMT_program);
		error++;
	}
	if (error) exit (EXIT_FAILURE);

	if (finite_out && no_time)
		sprintf (fmt, "%s%s%s%s%s", gmtdefs.d_format, gmtdefs.field_delimiter, gmtdefs.d_format, gmtdefs.field_delimiter, gmtdefs.d_format);
	else if (finite_out)
		sprintf (fmt, "%s%s%s%s%s%s%s", gmtdefs.d_format, gmtdefs.field_delimiter, gmtdefs.d_format, gmtdefs.field_delimiter, gmtdefs.d_format, gmtdefs.field_delimiter, gmtdefs.d_format);
	else
		sprintf (fmt, "%s%s%s%s%s%s%s%s%s", gmtdefs.d_format, gmtdefs.field_delimiter, gmtdefs.d_format, gmtdefs.field_delimiter, gmtdefs.d_format, gmtdefs.field_delimiter, gmtdefs.d_format, gmtdefs.field_delimiter, gmtdefs.d_format);
	last_sign = +1;
	for (i = 1; i < argc; i++) {
		if (!strncmp (argv[i], "-D", (size_t)2)) continue;
		if (!strncmp (argv[i], "-E", (size_t)2)) continue;
		if (!strncmp (argv[i], "-V", (size_t)2)) continue;
		if (!strncmp (argv[i], "-F", (size_t)2)) continue;
		if (!strncmp (argv[i], "-S", (size_t)2)) continue;
		if (!strncmp (argv[i], "-N", (size_t)2)) continue;
		if (!strncmp (argv[i], "-T", (size_t)2)) continue;

		if (argv[i][0] == '-' && argv[i][1] == '\0') {
			last_sign = -1;
			continue;
		}
		if (argv[i][0] == '+' && argv[i][1] == '\0') {
			last_sign = +1;
			continue;
		}

		if (GMT_access (argv[i], R_OK)) {	/* Not a readable file, is it a lon/lat/t0[/t1]/omega specification? */
			for (j = n_slash = 0; argv[i][j]; j++) if (argv[i][j] == '/') n_slash++;
			if ((n_slash + finite_in) < 3 || (n_slash + finite_in) > 4) {	/* No way it can be a online rotation, cry foul */
				fprintf (stderr, "%s: ERROR: Cannot read file %s\n", GMT_program, argv[i]);
				exit (EXIT_FAILURE);
			}
			else {	/* Tru to decode as a single rotation */

				j = sscanf (argv[i], "%lf/%lf/%lf/%lf/%lf", &lon, &lat, &t0, &t1, &angle);
				if (finite_in) angle = t1, t1 = 0.0;			/* Only 4 input values */
				if (n_slash == 2) angle = t0, t0 = 1.0, t1 = 0.0, no_time = TRUE;	/* Quick lon/lat/angle finite rotation, no time */
				if (t0 < t1) {
					fprintf (stderr, "%s: ERROR: Online rotation has t_start (%g) younger than t_stop (%g)\n", GMT_program, t0, t1);
					exit (EXIT_FAILURE);
				}
				if (angle == 0.0) {
					fprintf (stderr, "%s: ERROR: Online rotation has zero opening angle\n", GMT_program);
					exit (EXIT_FAILURE);
				}
				online_rot = TRUE;
			}
		}
		else
			online_rot = FALSE;

		if (first) {	/* First time loading a rotation model */
			if (online_rot) {
				n_a = 1;
				a = (struct EULER *) GMT_memory (VNULL, (size_t)1, sizeof (struct EULER), GMT_program);
				a[0].lon = lon;	a[0].lat = lat;
				a[0].t_start = t0;	a[0].t_stop = t1;
				a[0].duration = t0 - t1;
				a[0].omega = angle / a[0].duration;
				if (!finite_in) spotter_stages_to_finite (a, n_a, TRUE, TRUE);
			}
			else
				n_a = spotter_init (argv[i], &a, FALSE, finite_in, TRUE, &zero, gmtdefs.verbose);	/* Return finite rotations */
			zero = 0.0;
			if (last_sign == -1) {	/* Leading - sign, simply reverse the rotation angles */
				for (j = 0; j < n_a; j++) {
					a[j].omega = -a[j].omega;
					spotter_cov_of_inverse (&a[j], a[j].C);
				}
				last_sign = 1;
			}
			first = FALSE;
		}
		else {			/* For additional times, load a second model and add/subtract them */
			if (online_rot) {
				n_b = 1;
				b = (struct EULER *) GMT_memory (VNULL, (size_t)1, sizeof (struct EULER), GMT_program);
				b[0].lon = lon;	b[0].lat = lat;
				b[0].t_start = t0;	b[0].t_stop = t1;
				b[0].duration = t0 - t1;
				b[0].omega = angle / b[0].duration;
				if (!finite_in) spotter_stages_to_finite (b, n_b, TRUE, TRUE);
			}
			else
				n_b = spotter_init (argv[i], &b, FALSE, finite_in, TRUE, &zero, gmtdefs.verbose);	/* Return finite rotations */
			zero = 0.0;
			spotter_add_rotations (a, n_a, b, last_sign * n_b, &p, &n_p);			/* Add the two finite rotations sets, returns finite rotations in p */
			GMT_free ((void *)a);
			GMT_free ((void *)b);
			a = p;
			n_a = n_p;
		}
	}
	if (finite_out && no_time && header)
		printf ("#longitude%slatitude%sangle(deg)\n", gmtdefs.field_delimiter, gmtdefs.field_delimiter);
	else if (finite_out && header)	/* Easy, simply output what we've got following a header*/
		printf ("#longitude%slatitude%s%s%sangle(deg)\n", gmtdefs.field_delimiter, gmtdefs.field_delimiter, time_text[rate_or_rad], gmtdefs.field_delimiter);
	else if (finite_out)		/* Easy, simply output what we've got without a header */
		i = 0;	/* Do nothing here really */
	else {	/* Convert finite to stages before output */
		spotter_finite_to_stages (a, n_a, TRUE, TRUE);				/* To ensure we have the right kind of poles for output */
		if (header) printf ("#longitude%slatitude%s%s%s%s%sangle(deg)\n", gmtdefs.field_delimiter, gmtdefs.field_delimiter, start_text[rate_or_rad], gmtdefs.field_delimiter, end_text[rate_or_rad], gmtdefs.field_delimiter);
	}

	for (i = 0; i < n_a; i++) {
		if (transpose) a[i].omega = -a[i].omega;
		if ((south && a[i].lat > 0.0) || (north && a[i].lat < 0.0) || (!(south || north) && a[i].omega < 0.0))	/* flip to antipole */
			a[i].lat = -a[i].lat, a[i].lon += 180.0, a[i].omega = -a[i].omega;
		while (a[i].lon >= 360.0) a[i].lon -= 360.0;				/* Force geodetic longitude range */
		while (a[i].lon <= -180.0) a[i].lon += 360.0;				/* Force geodetic longitude range */
		while (geodetic && a[i].lon < 0.0) a[i].lon += 360.0;			/* Force geodetic longitude range */
		while (!geodetic && a[i].lon > 180.0) a[i].lon -= 360.0;		/* Force a geographic longitude range */
		if (reduce_angle) a[i].omega *= reduce_fact;
		if (finite_out && no_time)
			printf (fmt, a[i].lon, a[i].lat, a[i].omega * a[i].duration);
		else if (finite_out)
			printf (fmt, a[i].lon, a[i].lat, a[i].t_start, a[i].omega * a[i].duration);
		else
			printf (fmt, a[i].lon, a[i].lat, a[i].t_start, a[i].t_stop, a[i].omega * a[i].duration);
		if (a[i].has_cov) {
			double K[9];
			spotter_covar_to_record (&a[i], K);
			for (k = 0; k < 9; k++) {
				printf ("%s", gmtdefs.field_delimiter);
				printf (gmtdefs.d_format, K[k]);
			}
		}
		printf ("\n");
	}

	GMT_free ((void *)a);
	GMT_end (argc, argv);

	exit (EXIT_SUCCESS);
}
